engine/classes/Elgg/Filesystem/AdapterService.php

```diff
@@ -0,0 +1,96 @@
+<?php
+namespace Elgg\Filesystem;
+use Elgg\Filesystem\Adapter\Adapter;
+
+
+class AdapterService {
+	private $logger;
+	
+	/**
+	 *
+	 * @var string
+	 */
+	private $default;
+	
+	/**
+	 *
+	 * @var Adapter[]
+	 */
+	private $adapters;
+	
+	public function __construct(\Elgg\Logger $logger) {
+		$this->logger = $logger;
+	}
+	
+	public function set($name, Adapter $adapter, $default = false) {
+		if ($this->has($name)) {
+			throw new \UnexpectedValueException("Filesystem adapter \"$name\" already exists.");
+		}
+		$this->adapters[$name] = $adapter;
+		
+		if ($default) {
+			$this->setDefault($name);
+		}
+		return true;
+	}
+	
+	/**
+	 * 
+	 * @param type $name
+	 * @return Adapter
+	 * @throws \InvalidArgumentException
+	 */
+	public function get($name) {
+		if (!$this->has($name)) {
+			throw new \InvalidArgumentException("Filesystem Adapter \"$name\" does not exist.");
+		}
+		
+		return $this->adapters[$name];
+	}
+	
+	public function has($name) {
+		return isset($this->adapters[$name]);
+	}
+	
+	public function remove($name) {
+		// throw if doesn't exist
+		$this->get($name);
+		unset($this->adapters[$name]);
+		return true;
+	}
+	
+	public function listAdapters() {
+		return array_keys($this->adapters);
+	}
+	
+	public function setDefault($name) {
```

Comment:
The default adapter isn't available until after `init, system`, so we can't set it in the services as a constructor argument. I'm leaning toward removing the default stuff from this class and requiring explicit calls for adapter types. The default filestore concept looks like it was to keep from crashing if the filestore info changed, so I'm not convinced it's the right solution.